### PR TITLE
Add Image Dataloader for PASCAL VOC and CIFAR 10 dataset.

### PR DESCRIPTION
**Pull request by kartikdutt18**
Hey everyone,
This is is a WIP of image data loader for this repo.

TO DO :
- [x] Load labels into the data loader.
- [x] Add Resize function to reshape loaded images (by 31st).
       (maybe another PR or in this one, kindly let me know).
- [x] Load Images. (by 1st)
- [x] Create subset of the dataset to upload on mlpack.org for testing. (By 2nd)
- [x] Write tests for the same. (By 2nd)
- [x] Repeat for classification of imagenet
       (Since the code is similar it will only take a day or two after completing PASCAL VOC).

### CODE REVIEW

**Comment by KimSangYeon-DGU** [comment]
Hi @kartikdutt18. Regarding `Add Resize function to reshape loaded images`, if this topic is about the discussion we had in the Community Bonding period, let's put this off. Although I suggested that before, I think the feature to set the size of an image when loading is not vital for this summer work. In addition, the resizing feature is more related to image utilities, not data loader.

Please let me know what you think

**Comment by kartikdutt18** [comment]
Hey @KimSangYeon-DGU, This is something I missed while making the proposal, The images in PASCAL dataset have different sizes, I can preprocess the images in python and we can store that dataset or We can add a resize feature that uses Bilinear interpolation from mlpack to resize images. Kindly let me know what you think.

**Comment by KimSangYeon-DGU** [comment]
@kartikdutt18 Ok, then, let's do that :) I don't mind if you do this work on either this PR or another.

**Comment by kartikdutt18** [comment]
> @kartikdutt18 Ok, then, let's do that :) I don't mind if you do this work on either this PR or another.

Great, I'll do it in this PR first, if get a bit large I can open another one with just that. Thanks a lot.

**Comment by KimSangYeon-DGU** [comment]
Yes, the best is to implement and use resizing feature in mlpack :) BTW, please let me know if this work turns into a burden for the summer work. We can adjust by going another way as you said above comment.

**Comment by kartikdutt18** [comment]
> Yes, the best is to implement and use resizing feature in mlpack :) BTW, please let me know if this work turns into a burden for the summer work. We can adjust by going another way as you said above comment.

Thanks a lot, If things go well I should have the resize function ready by tomorrow or the day after. Also, I think it'll be nice to have that feature.

**Comment by KimSangYeon-DGU** [comment]
Or we can use OpenCV as well

**Comment by kartikdutt18** [comment]
> Or we can use OpenCV as well

Right, That would be much easier. I think have the conversion code for opencv matrix to armadillo so that should be easy to do. Thanks a lot for the amazing suggestion.

**Comment by KimSangYeon-DGU** [comment]
Yes, and as you know, if we use OpenCV, it can do  that in `mlpack/examples`

**Comment by kartikdutt18** [comment]
Hey @KimSangYeon-DGU, I completed the main function for PASCAL Dataloader i.e. `LoadObjectDetectionDataset`. Tested with couple of images and it works fine. I have also added a resize function and augmentation class. In augmentation class only resize is supported for now. Maybe we can open issue if anyone wants to implement some augmentation they can. I haven't used open cv yet because that was much slower than mlpack since we had to iterate over each column however here we pass a batch of images.
Things left to do are mentioned below :
1. Clean up.
2. Writing tests.
3. Downloading Pascal VOC dataset from constructor and calling LoadObjectDetectionDataset from constructor.
4. Writing Documentation for both Object Detection Dataloader and Augmentation class (Might be better to have a separate PR).
5. Uploading a subset of Dataset to mlpack.org for testing since its not practical to download whole dataset for testing (Roughly 2GB).

**Comment by kartikdutt18** [comment]
Hey @KimSangYeon-DGU, just another update. I have tests for both the data loader as well as augmentation class. I have created a subset that we would can upload to mlpack.org that can be used for testing. Only one thing is left i.e. Calling function through the constructor. If this makes sense I can add that in the next commit.
Regards.

**Comment by KimSangYeon-DGU** [comment]
Nice @kartikdutt18! Since #12 got merged, can you re-base this branch? it helps to review.

File: augmentation/augmentation_impl.hpp [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+  // We will use mlpack's bilinear interpolation layer to
+  // resize the input.
+  mlpack::ann::BilinearInterpolation<DatasetType, DatasetType> resizeLayer(
+      datapointWidth, datapointHeight, outputWidth, outputHeight,
+      datapointDepth);
+
+  // Not sure how to avoid a copy here.
+  DatasetType output;
+  resizeLayer.Forward(dataset, output);
+  dataset = output;
```
Comment (kartikdutt18) [inline augmentation/augmentation_impl.hpp:96]: I have resolved all conflicts. The one thing I am unsure of is how to avoid a copy here. For very large datasets resizing will simply not be possible incase the program quits as they run out of RAM.

**Comment by kartikdutt18** [comment]
Also one thing that we could add maybe later (or open an issue) is addition of script to convert csv to xml files.

File: augmentation/augmentation_impl.hpp (hunk 1 above)
Comment (KimSangYeon-DGU) [reply augmentation/augmentation_impl.hpp:96]: How about use this?
```suggestion
  // Not sure how to avoid a copy here.
  DatasetType output;
  resizeLayer.Forward(dataset, output);
  dataset = std::(output);
```

Comment (kartikdutt18) [reply augmentation/augmentation_impl.hpp:96]: Ahh makes sense. Awesome, Thanks a lot for this.

Comment (kartikdutt18) [reply augmentation/augmentation_impl.hpp:96]: One more thing I was planning to change was to use fields in labels. So when labels are passed in preprocessor function for YOLO model it would easier than doing look ups. Currently each labels has a different image, we could use a field vector and store multiple bounding boxes in one column rather than having a separate column for each bounding box. Let me know if this makes sense.

Comment (KimSangYeon-DGU) [reply augmentation/augmentation_impl.hpp:96]: > Currently each labels has a different image, we could use a field vector and store multiple bounding boxes in one column rather than having a separate column for each bounding box

Can you do that as a PoC? and then, let's see which one is better. I need to figure out how the field vector works.

At first, I thought there will be a problem in using linear algebra operations when we put multiple bounding boxes into one column. However, as you said, each label has a different shape of image, so we could do that.

File: augmentation/augmentation.hpp [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+ * augmentation.Transform(dataloader.TrainFeatures);
+ * @endcode
+ * 
+ * @tparam DatasetType Datatype on which augmentation will be done.
+ */
+template<typename DatasetType = arma::mat>
+class Augmentation
+{
+ public:
+  //! Create the augmenation class object.
```
Comment (KimSangYeon-DGU) [inline augmentation/augmentation.hpp:34]:
```suggestion
  //! Create the augmentation class object.
```

File: augmentation/augmentation.hpp [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+  Augmentation();
+
+  /**
+   * Constructor for augmentation class.
+   *
+   * @param augmentations List of strings containing one of the supported
+   *                     augmentation.
```
Comment (KimSangYeon-DGU) [inline augmentation/augmentation.hpp:41]:
```suggestion
   * @param augmentations List of strings containing one of the supported
   *                     augmentations.
```

File: augmentation/augmentation_impl.hpp [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+
+template<typename DatasetType>
+Augmentation<DatasetType>::Augmentation(
+    const std::vector<std::string>& augmentations,
+    const double augmentationProbability) :
+    augmentations(augmentations),
+    augmentationProbability(augmentationProbability)
+{
+  // Sort the vector to place resize parameter to the front of the string.
+  // This prevents constant look ups for resize.
```
Comment (KimSangYeon-DGU) [inline augmentation/augmentation_impl.hpp:34]: I think it should be `lookups`.

```suggestion
  // This prevents constant lookups for resize.
```

File: augmentation/augmentation_impl.hpp [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+  sort(this->augmentations.begin(), this->augmentations.end(), [](
+      std::string& str1, std::string& str2)
+          {
+            return str1.find("resize") != std::string::npos;
+          });
```
Comment (KimSangYeon-DGU) [inline augmentation/augmentation_impl.hpp:39]: Can you check this style?

File: augmentation/augmentation.hpp [hunk 6]
```diff
@@ -0,0 +1,174 @@
...
+   *                     are extracted.
+   */
+  void GetResizeParam(size_t& outWidth,
+                      size_t& outHeight,
+                      const std::string& augmentation)
+  {
+    if (!HasResizeParam())
+    {
+      return;
+    }
```
Comment (KimSangYeon-DGU) [inline augmentation/augmentation.hpp:122]: Likewise
```suggestion
    if (!HasResizeParam())
      return;
```

File: augmentation/augmentation.hpp [hunk 7]
```diff
@@ -0,0 +1,174 @@
...
+   * Function to determine if augmentation has Resize function.
+   * @param augmentation Optional argument to check if a string has
+   *                     resize substring.
+   */
+  bool HasResizeParam(const std::string& augmentation = "")
+  {
+    if (augmentation.length())
+    {
+      return augmentation.find("resize") != std::string::npos;
+    }
```
Comment (KimSangYeon-DGU) [inline augmentation/augmentation.hpp:100]: According the [Brace Placement in DesignGuidelines](https://github.com/mlpack/mlpack/wiki/DesignGuidelines#brace-placement), let's remove the braces :)
```suggestion
    if (augmentation.length())
      return augmentation.find("resize") != std::string::npos;
```

File: augmentation/augmentation_impl.hpp [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+    DatasetType& dataset,
+    const size_t datapointWidth,
+    const size_t datapointHeight,
+    const size_t datapointDepth,
+    const std::string& augmentation)
+{
+  if (!this->HasResizeParam(augmentation))
+  {
+    return;
+  }
```
Comment (KimSangYeon-DGU) [inline augmentation/augmentation_impl.hpp:80]: Likewise
```suggestion
  if (!this->HasResizeParam(augmentation))
    return;
```

File: dataloader/dataloader_impl.hpp [hunk 9]
```diff
@@ -145,4 +145,137 @@ template<
...
+  Utils::ListDir(pathToAnnotations, annotationsDirectory, absolutePath);
+  Utils::ListDir(pathToImages, imagesDirectory, absolutePath);
+
+  // Create a map for labels and corresponding class name.
+  // This provides faster access to class labels.
+  std::unordered_map<std::string, size_t> classMap;
+  for (size_t i = 0; i < classes.size(); i++)
+  {
+    classMap.insert(std::make_pair(classes[i], i));
+  }
```
Comment (KimSangYeon-DGU) [inline dataloader/dataloader_impl.hpp:185]:
```suggestion
  for (size_t i = 0; i < classes.size(); i++)
    classMap.insert(std::make_pair(classes[i], i));
```

File: dataloader/dataloader.hpp [hunk 10]
```diff
@@ -117,6 +121,60 @@ class DataLoader
...
+   * 1. Each XML file should be wrapped in annotation tag.
+   * 2. Filename of image in images folder will be depicted by filename tag.
+   * 3. Object tag depicting characteristics of bounding box.
+   * 4. Each object tag should contain name tag i.e. class of the object.
+   * 5. Each object tag should contain bndbox tag containing xmin, ymin, xmax, ymax.
+   *
+   * NOTE : Labels are assigned using classes vector. Set verbose to 1 to print labels
+   * and their corresponding class.
+   *
+   * @param pathToAnnotations Path to the folder containg xml type annotation files.
```
Comment (KimSangYeon-DGU) [inline dataloader/dataloader.hpp:138]:
```suggestion
   * @param pathToAnnotations Path to the folder containing XML type annotation files.
```

File: dataloader/dataloader.hpp [hunk 11]
```diff
@@ -117,6 +121,60 @@ class DataLoader
                    std::vector<std::string>(),
                const double augmentationProbability = 0.2);
 
+  /**
+   * Loads object detection dataset. It requires a single annotation file in xml format.
```
Comment (KimSangYeon-DGU) [inline dataloader/dataloader.hpp:125]:
```suggestion
   * Loads object detection dataset. It requires a single annotation file in XML format.
```

File: dataloader/dataloader_impl.hpp [hunk 12]
```diff
@@ -145,4 +145,137 @@ template<
...
+
+  // Map to insert values in a column vector.
+  std::unordered_map<std::string, size_t> indexMap;
+  indexMap.insert(std::make_pair(classNameXMLTag, 0));
+  indexMap.insert(std::make_pair(x1XMLTag, 1));
+  indexMap.insert(std::make_pair(y1XMLTag, 2));
+  indexMap.insert(std::make_pair(x2XMLTag, 3));
+  indexMap.insert(std::make_pair(y2XMLTag, 4));
+
+  // Read the xml file.
```
Comment (KimSangYeon-DGU) [inline dataloader/dataloader_impl.hpp:195]:
```suggestion
  // Read the XML file.
```

File: dataloader/dataloader_impl.hpp [hunk 13]
```diff
@@ -145,4 +145,137 @@ template<
...
+  for (boost::filesystem::path annotationFile : annotationsDirectory)
+  {
+    if (annotationFile.string().length() <= 3 ||
+        annotationFile.string().substr(
+            annotationFile.string().length() - 3) != "xml")
+    {
+      continue;
+    }
+    // Read the xml file.
```
Comment (KimSangYeon-DGU) [inline dataloader/dataloader_impl.hpp:204]:
```suggestion
    // Read the XML file.
```

File: dataloader/dataloader_impl.hpp [hunk 14]
```diff
@@ -145,4 +145,137 @@ template<
...
+    boost::property_tree::ptree xmlFile;
+    boost::property_tree::read_xml(annotationFile.string(), xmlFile);
+
+    // Get annotation from xml file.
```
Comment (KimSangYeon-DGU) [inline dataloader/dataloader_impl.hpp:208]:
```suggestion
    // Get annotation from XML file.
```

File: dataloader/dataloader_impl.hpp [hunk 15]
```diff
@@ -145,4 +145,137 @@ template<
...
+    boost::property_tree::ptree annotation = xmlFile.get_child(baseXMLTag);
+
+    // Read properties inside annotation file.
+    // Get image name.
+    std::string imgName = annotation.get_child(imageNameXMLTag).data();
+
+    // If image doesn't exist then skip the current xml file.
```
Comment (KimSangYeon-DGU) [inline dataloader/dataloader_impl.hpp:215]:
```suggestion
    // If image doesn't exist then skip the current XML file.
```

File: augmentation/augmentation_impl.hpp (hunk 1 above)
Comment (kartikdutt18) [reply augmentation/augmentation_impl.hpp:96]: Sure, I'll make the changes now.

File: dataloader/dataloader_impl.hpp (hunk 15 above)
Comment (kartikdutt18) [reply dataloader/dataloader_impl.hpp:215]: Sorry about the style changes, Will try to be more consistent with wording. Thanks a lot for the review. Making the changes now.

File: augmentation/augmentation_impl.hpp (hunk 1 above)
Comment (kartikdutt18) [reply augmentation/augmentation_impl.hpp]: >  Repeat for classification of imagenet
(Since the code is similar it will only take a day or two after completing PASCAL VOC).

I think this PR is almost ready, I am working on the last to-do now. Will complete it by tomorrow so that we can have the same functionality as FlowFromDirectory (keras) for image datasets. The Pascal dataloader is complete. It takes roughly 18 or more minutes to load all the images though. About the field vector I'll create another branch same as this one and post this link here. If that is better I can simply replace the commits.

Comment (kartikdutt18) [reply augmentation/augmentation_impl.hpp]: Added image dataloader with flow from directory functionality. We can now support voc-detection, voc-classification and cifar-10 / 100. I'll set up the constructor and do the clean up as well. The build error in linux / windows now is because they have .DS_Store file. I'll add a check for that as well. Currently just manually checking for the file.

**Comment by kartikdutt18** [comment]
Hey @KimSangYeon-DGU, mlpack's internal split (for labels and features) doesn't work for any other armadillo type other than rowvec and mat type respectively. Should I open a PR to template it or implement it locally here. We would need the validation set for PASCAL after training YOLO because test set of PASCAL is on private evaluation server so they don't host it publicly (unlike the training set). Kindly let me know what you think.
Regards.

**Comment by KimSangYeon-DGU** [comment]
Per discussion and given the timeline, we'll implement the split method locally for internal use for the time being and open an issue about that the mlpack's `Split()` method needs to use template to work with other Armadillo data types. If the Split() is changed to a template method, we will use it in the future.

**Comment by kartikdutt18** [comment]
Agreed, opened mlpack/mlpack#2443 for future discussion.

File: augmentation/augmentation.hpp [hunk 16]
```diff
@@ -0,0 +1,172 @@
...
+ * For the list of supported augmentation, take a look at our wiki page.
+ *
+ * @code
+ * Augmentation<> augmentation({"horizontal-flip", "resize = (224, 224)"}, 0.2);
+ * augmentation.Transform(dataloader.TrainFeatures);
+ * @endcode
+ * 
+ * @tparam DatasetType Datatype on which augmentation will be done.
+ */
+template<typename DatasetType = arma::mat>
```
Comment (zoq) [inline augmentation/augmentation.hpp:30]: Seems like the class template parameter isn't needed, if we add a template for the method.

File: augmentation/augmentation.hpp [hunk 17]
```diff
@@ -0,0 +1,172 @@
...
+{
+ public:
+  //! Create the augmentation class object.
+  Augmentation();
+
+  /**
+   * Constructor for augmentation class.
+   *
+   * @param augmentations List of strings containing one of the supported
+   *                     augmentations.
```
Comment (zoq) [inline augmentation/augmentation.hpp:41]: Either this should be tabbed twice (4 spaces) or should align with the first line of the parameter description.

File: augmentation/augmentation.hpp (hunk 16 above)
Comment (kartikdutt18) [reply augmentation/augmentation.hpp:30]: Ahh, I was doing that earlier but it gave me a strange error with the map [here](https://github.com/mlpack/models/pull/13/files/e97e616a80ee81b2afb8ee746f18663a13c208fc#diff-d8125dd9142e99fd90c585af5f57daa4R161), 
```
std::unordered_map<std::string, void(*)(DatasetType&,
       size_t, size_t, size_t, std::string&)> augmentationMap;
```
So I decided to added a template for the class rather than the method.

File: augmentation/augmentation.hpp (hunk 17 above)
Comment (kartikdutt18) [reply augmentation/augmentation.hpp:41]: Sure, will make the change here and elsewhere too. Thanks a lot.

File: augmentation/augmentation.hpp (hunk 16 above)
Comment (zoq) [reply augmentation/augmentation.hpp:30]: But looks like the container is used in a single method, so I could be instantiated locally as well? I think removing the class template would make it easier for a user to use the augmentation feature.

Comment (KimSangYeon-DGU) [reply augmentation/augmentation.hpp:30]: > I think removing the class template would make it easier for a user to use the augmentation feature.

Agreed, @kartikdutt18 can you try do that in #18?

Comment (kartikdutt18) [reply augmentation/augmentation.hpp:30]: Ahh, makes sense. Will make the change in #18.

**Comment by kartikdutt18** [comment]
This PR is migrated to #18. Thanks a lot.
